fix: Don't attach missing or substituted fonts on submit

cilevitz opened the pull request:
### What was the problem/requirement? (What/Why)
The submitter was collecting all font files in a project, including substituted fonts resulting from After Effects loading a project with missing or previous substituted fonts. Render tasks proceeded using the substituted fonts, producing incorrect output despite the tasks being marked as Succeeded.

### What was the solution? (How)
We filter out any missing or substituted fonts from the list of used font files that gets processed when collecting job attachments. We also adjusted `call_aerender.py` so that it errors out if it detects the warning `Project has missing fonts` in the render log.

### What is the impact of this change?
Users can use any custom fonts and cloud fonts they'd like, and all chosen fonts will render correctly on the farm. The user is alerted to any missing or substituted fonts before opening the submitter. Missing fonts will cause a render task to stop and be marked "Failed".

### How was this change tested?
Testing was performed with SMF hosts running After Effects 24.0 and 25.1 to verify correct `aerender` output of a project containing custom and cloud fonts. The farm render output was compared to local renders from a machine with those fonts installed, and the farm output is as expected. The log output parser in `call_aerender.py` stops and fails the render task if a missing font warning is seen.

### Was this change documented?
The README.md was updated with notes about missing and substituted fonts.

### Is this a breaking change?
No.

----

*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

eherozhao committed:
feat:Dockable submitter (#63)

* feat: dockable submitter using aerender

Signed-off-by: Hao Zhao <36955586+eherozhao@users.noreply.github.com>

eherozhao committed:
docs: update README, CONTRIBUTING and DEVELOPMENT doc (#79)

* docs: update README, CONTRIBUTING and DEVELOPMENT doc
---------
Signed-off-by: Hao Zhao <36955586+eherozhao@users.noreply.github.com>

eherozhao committed:
fix: provide a way for CMF customers with aerender exectuable env option (#80)

* fix: provide a way for CMF customers with aerender exectuable env options

---------

Signed-off-by: Hao Zhao <36955586+eherozhao@users.noreply.github.com>

eherozhao committed:
chore: delete unuseful log (#82)

* chore: remove unused logging

---------

Signed-off-by: Hao Zhao <36955586+eherozhao@users.noreply.github.com>

eherozhao committed:
fix: use shell executable to call deadline cli for macOS. (#83)

* fix: update the MacOS submit process to use bash as to avoid deadline path issue

---------

Signed-off-by: Hao Zhao <36955586+eherozhao@users.noreply.github.com>

eherozhao committed:
chore: resolve merge conflict

Signed-off-by: Hao Zhao <36955586+eherozhao@users.noreply.github.com>

eherozhao committed:
chore: remove unused code (#84)

* chore: remove unused code

---------

Signed-off-by: Hao Zhao <36955586+eherozhao@users.noreply.github.com>

eherozhao committed:
feat: add hostRequirement for AE job template (#85)

feat: add hostRequirement windows for AE job template
---------

Signed-off-by: Hao Zhao <36955586+eherozhao@users.noreply.github.com>

viknith committed:
fix: Remove CondaPackages parameter override to restore default param... (#86)

* fix: Remove CondaPackages parameter override to restore default parameter set in queue environment
---------
Signed-off-by: Vikram Nithyanandam <143119182+viknith@users.noreply.github.com>

eherozhao committed:
fix: replace hard-coded template.yaml with code generated json file (#87)

Signed-off-by: Hao Zhao <36955586+eherozhao@users.noreply.github.com>

eherozhao committed:
fix: Use JSON data structure to change job template parameters (#88)

Signed-off-by: Hao Zhao <36955586+eherozhao@users.noreply.github.com>

Commit:
feat: Add custom font support by detecting fonts used in a composition and submitting them as job attachments

Commit:
feat: Show popup if missing fonts are detected when collecting job attachments
feat: Multi-level logging output in font_manager.py
docs: Update README
fix: Direct invocation of python font_manager.py install
fix: Update docstrings
fix: Loop index names
fix: f-string formatting

eherozhao committed:
feat: image sequences chunking (#93)

Signed-off-by: Hao Zhao <36955586+eherozhao@users.noreply.github.com>

Commit:
feat: Use app.project.usedFonts when available for font collection
fix: Scan all TextLayer keys for font changes
fix: Remove .ttc support
docs: Update README

eherozhao committed:
fix: add a release bump yaml (#97)

Signed-off-by: Hao Zhao <36955586+eherozhao@users.noreply.github.com>

eherozhao committed:
fix: replace %20 with space for Windows os (#98)

Signed-off-by: Hao Zhao <36955586+eherozhao@users.noreply.github.com>

Commit:
fix: Update warning dialog message for fonts without a fontLocation
docs: Update documentation with font installation steps

eherozhao committed:
feat: add xml file for submitter installer (#96)

Signed-off-by: Hao Zhao <36955586+eherozhao@users.noreply.github.com

Commit:
Merge remote-tracking branch 'remotes/origin/feature_dockable_submitter' into feature_font_support

Commit:
chore: Update dist

joel-wong-aws committed:
chore: update GitHub issue templates (#99)

Signed-off-by: Joel Wong <127782171+joel-wong-aws@users.noreply.github.com>

Commit:
chore: Adjust spaces and line spacing
chore: Update CODEOWNERS

Commit:
feat: Check Python version with getPythonExecutable()
fix: getFontPaths() error alert
chore: Refactor get_font_name()

Commit:
feat: Show alert on error running get_user_fonts.py

Commit:
chore: Remove blank line

Signed-off-by: Daniel Cilevitz <daniel@redtoast.ca>

cilevitz committed:
Merge branch 'feature_dockable_submitter' into feature_font_support

Commit:
chore: Update requirements.txt with fonttools library
fix: Update README.md with fonttools installation
fix: Add exception types to exception handlers

Signed-off-by: Daniel Cilevitz <daniel@redtoast.ca>

Commit:
fix: Don't attach missing or substituted fonts on submit
fix: Show error dialog if fatal error when finding fonts
fix: Fail render tasks with an error if missing fonts at render time

Commit:
Merge branch 'mainline' into feature_font_support

cilevitz committed:
Merge branch 'aws-deadline:mainline' into feature_font_support

Signed-off-by: Daniel Cilevitz <daniel@redtoast.ca>

cilevitz committed:
Merge branch 'aws-deadline:mainline' into feature_font_support

Commit:
Merge branch 'feature_font_support' of https://github.com/redtoastproductions/deadline-cloud-for-after-effects into feature_font_support

Commit:
docs: Update README.md about missing and substituted fonts

Signed-off-by: Daniel Cilevitz <daniel@redtoast.ca>

eherozhao commented on dist/DeadlineCloudSubmitter_Assets/JobTemplate/scripts/call_aerender.py:
```diff
@@ -91,6 +91,11 @@ def main():
                 print(f"After Effects Warning: {line.strip()}", file=sys.stderr)
 
+                # Some warnings should be treated as errors, i.e. a fatal warning
+                if "Project has missing fonts" in line:
```
Do we need to make sure in the submitter side, we have enough warning to customers that the submitted project with missing fonts will fail?

eherozhao replied:
I am trying to understand this behavior from customer perspective:
1. If I am a customer wanting to submit a job with missing custom font and I don't really care about the missing fonts, the render should succeeded.
2. If I am a customer wanting to submit a job with missing custom font and I want to make the change to update them before the job is rendered in the Deadline Cloud side.
In this way, I think we can have a warning pop up for customers that "are you sure you are going to submit a job with missing fonts?" with a yes and no ask. If yes, our job attachment should auto detect the replaced fonts with the replaced font name (not the one shown in the application)" If no, we just cancel the job submission.

I don't think we always fail customers missing fonts in the worker with no ask is a good customer experience, especially the worker has to run all steps before this failure, which seems wasting their money.

cilevitz replied:
We added a confirmation dialog when submitting a composition containing missing or substituted fonts.

eherozhao commented on dist/DeadlineCloudSubmitter_Assets/JobTemplate/scripts/get_user_fonts.py:
```diff
@@ -72,7 +72,7 @@ def get_font(font_path):
     try:
         raw_table = {i:str(names_table[i]) for i in range(0, len(names_table))}
-    except Exception:
+    except Exception as e:
```
q: if we don't print out error in the block, do we still need `as e`? I would suggest us either print the error out or delete `as e`.

Also, are we able to print out the python results from subprocess to the main submitter output to surface to customers?

cilevitz replied:
Font scan errors from Python are now surfaced in the AE UI.

viknith commented on src/submission/JobTemplateHelper.jsx:
```diff
@@ -141,10 +141,24 @@ function findJobAttachments(rootComp) {
+        for (var i = 0; i < fontsInProject.length; i++) {
+            var attachFont = true;
+            // If the font's fontPostScriptName contains a missing or substituted font name, don't attach it
+            if (app.fonts.missingOrSubstitutedFonts.toString().indexOf(fontsInProject[i].fontPostScriptName) !== -1) {
```
Wouldn't the alternative behavior be to just name the files correctly with the fonts that will actually be used in place of the missing fonts? IIRC, that was the confusing part, since the substituted font files were being renamed to fonts that actually won't be included. It may be possible that customers are aware that their composition is missing the correct fonts for now, but don't want to modify the composition and allow Adobe to substitute in some default fonts for now and just get a job rendering for now. 

This behavior could be toggle-able in case customers still want to submit the job with substituted fonts and want them to pass, and at the same time we can avoid misnaming files. Might be out of scope of this PR, bu it is worth discussing.

viknith replied:
Essentially, warnings are good, but I'm concerned about the job failing due to substituted fonts, and we should have a checkbox for customers on whether customers want substituted font jobs to pass or fail.

cilevitz replied:
Missing or substituted fonts are now added to job attachments with the name of their substitution, e.g. `CyTextBold` substituted to `ArialMT` gives an attachment named `ArialMT`.

Missing fonts will not cause a job to fail on the farm, but will produce incorrect render output.

The user is warned when clicking Submit that they should install missing or substituted fonts to ensure correct render output.

eherozhao replied:
If I am using a unsupported ttc font type, what will it behave? I didn't see the notification that the font was substituted to the default font so I think it might be a missing part when the font was not supported (but not missing). Wanna hear about your thoughts on this.

cilevitz replied:
Submitting a project from Mac with a TTC font unsupported by Windows (and no other missing or substituted fonts at submit time) will result in no warnings from the submitter. The font will not be installed by font_manager.py on a Windows render node, as unsupported fonts cannot be installed. aerender will warn about missing fonts at render time. Render output will have substituted fonts.

viknith commented on src/submission/JobTemplateHelper.jsx:
```diff
@@ -141,10 +141,24 @@ function findJobAttachments(rootComp) {
+        // Build a new array containing only attachable fonts.
+        var fontsInProjectFiltered = [];
+        for (var i = 0; i < fontsInProject.length; i++) {
+            var attachFont = true;
```
Also this code can be simplified to an if/else structure, no need for extra attachFont boolean:
```
if (app.fonts.missingOrSubstitutedFonts.toString().indexOf(fontsInProject[i].fontPostScriptName) !== -1) {
   logger.warning("Not attaching missing or substituted font: " + fontsInProject[i].fontPostScriptName, jobTemplateHelperFile);
} else {
    fontsInProjectFiltered.push(fontsInProject[i]);
}
```

cilevitz replied:
Resolved in c9c25ca

viknith commented:
**On Windows and MacOS for AE 24.6.4 and AE 25.1, I was able to confirm that the error messages popped up correctly:**

<img width="408" alt="Screenshot 2025-02-02 at 3 30 50 PM" src="https://github.com/user-attachments/assets/5ceb4635-5e82-4d19-b1a9-fc269d8edda9" />
<img width="322" alt="Screenshot 2025-02-02 at 3 32 26 PM" src="https://github.com/user-attachments/assets/73126b3c-cf61-4696-98be-c2b53b9c3a0e" />

**And that the job attachments did not include substituted files with the missing font names**

![Screenshot 2025-02-02 at 2 45 00 PM](https://github.com/user-attachments/assets/ffef345a-f576-4310-a095-a7cc76291e13)
![Screenshot 2025-02-02 at 3 22 05 PM](https://github.com/user-attachments/assets/37193131-dfd6-4d75-beb1-8b3917b69ba8)

And here was the job submission results when they were submitted, all failing.
<img width="1125" alt="Screenshot 2025-02-02 at 3 34 53 PM" src="https://github.com/user-attachments/assets/0d879196-de75-45a3-b63a-fdd3d727ef80" />
<img width="1130" alt="Screenshot 2025-02-02 at 3 49 22 PM" src="https://github.com/user-attachments/assets/7c41a8c1-5402-4411-9018-5832a19ec0fc" />

Seems to be all in working order! Just need to address some of the comments from above, I think the one from Hao about letting customers know that their job will likely fail due to the missing font is important.

eherozhao reviewed (changes requested):
Do you mind also fixing the SonarQube and signing? Thanks!

Commit:
Merge branch 'mainline' into feature_font_support

Commit:
feat: Show errors from get_user_fonts.py in the UI
fix: Attach substituted fonts by their substituted font name
fix: Warn instead of error if font files are in use when being copied or removed
fix: Warn instead of error for "Project has missing fonts"

Signed-off-by: Daniel Cilevitz <daniel@redtoast.ca>

Commit:
fix: Adjust variable declarations

Commit:
chore: Refactor getFontNameOverride, get_fonts

Signed-off-by: Daniel Cilevitz <daniel@redtoast.ca>

cilevitz committed:
Merge branch 'aws-deadline:mainline' into feature_font_support

cilevitz commented:
SonarQube issues have been resolved.

Commit:
feat: Add confirmation dialog for submitting a composition containing missing or substituted fonts

Signed-off-by: Daniel Cilevitz <daniel@redtoast.ca>

Commit:
Merge branch 'feature_font_support' of https://github.com/redtoastproductions/deadline-cloud-for-after-effects into feature_font_support

eherozhao commented on README.md:
```diff
@@ -76,11 +76,11 @@ The submitter includes a folder `DeadlineCloudSubmitter_Assets` and a file `Dead
 Windows bitmap fonts (`.fon`) are only supported on Windows machines.
 
-If fonts are missing at render time, first check that they're installed (on the system or your user), and then check they're being included in the job attachments tab in the submitter.
+If fonts are missing or substituted at render time, then render tasks will fail. You must add any required fonts to your system or your user profile before opening a project that needs them. First check that the missing fonts are installed (on the system or your user), and then check they're being included in the job attachments tab in the submitter.
```
I thought we wouldn't fail the render job if the font was substituted if I understood the implementation correctly. Please confirm it.

cilevitz replied:
Documentation has been updated in ea25a1f to indicate a choice to submit a job with missing or substituted fonts, even though the render output will not be correct.

viknith commented on dist/DeadlineCloudSubmitter_Assets/JobTemplate/scripts/font_manager.py:
```diff
@@ -177,13 +184,24 @@ def uninstall_font(src_path, scope=INSTALL_SCOPE_USER):
-            winreg.DeleteValue(key, fontname)
+        
+        try:
+            with winreg.OpenKey(registry_scope, FONTS_REG_PATH, 0, access= winreg.KEY_SET_VALUE) as key:
```
Nit: `access= winreg.KEY_SET_VALUE` has extra space after equal sign

cilevitz replied:
Resolved in ea25a1f

viknith commented on src/submission/JobTemplateHelper.jsx:
```diff
@@ -348,9 +407,43 @@ function createFontFilename(fontLocation, fontPostScriptName) {
+ * Collects metadata from a given font and determines if it needs an override name.
+ * @return Object containing the override name or an error
+ **/
+function getFontNameOverride(fontObject, fontPostScriptName, fontLocation, oldLocation) {
```
nit: oldLocation is an unused variable

cilevitz replied:
Resolved in ea25a1f

viknith commented on src/submission/JobTemplateHelper.jsx:
```diff
@@ -310,11 +344,36 @@ function getLocationForFont(fontPostScriptName) {
+    var fontPostScriptName = null;
+    try {
+        // Get user-installed fonts
+        var fontPaths = getFontPaths();
```
Not a blocker for this PR, but we should cache the output of getFontPaths because the output is the same each time

cilevitz replied:
fontPaths are now cached in ea25a1f

viknith commented on src/submission/SubmitBundle.jsx:
```diff
@@ -92,6 +92,17 @@ function SubmitSelection(selection, framesPerTask) {
     var compName = dcUtil.removeIllegalCharacters(rqi.comp.name);
 
+    if (app.fonts.missingOrSubstitutedFonts != "") {
+        var submitConfirmation = confirm(
```
Just tried submitting a job that had no missing or substituted fonts, all using just Amazon Ember and the GUI submitter CLI command was not reached and the submitter fails silently.

The root caused is happening around these lines of code.

Most likely it's because this variable is never created and set to true if there are no missing fonts. As a result, it defaults to false, returns, and then there's not notification to the customer. This needs to be set to default true initially and then have the if-check with confirmation. Requires code fix.

viknith replied:
Tried to nest the second if statement under the first with missing fonts but that didn't work, AE ends up loading and then crashes from not responding. This is AE 25.1 on MacOS btw

cilevitz replied:
Resolved in ea25a1f

eherozhao commented on src/submission/JobTemplateHelper.jsx:
```diff
@@ -140,46 +140,80 @@ function findJobAttachments(rootComp) {
+        }
+        // Build a new array containing only attachable fonts.
+        var fontsInProjectFiltered = [];
+        for (i = 0; i < fontsInProject.length; i++) {
```
nit: do we need it to be `var i = 0` here?

cilevitz replied:
`i` is reused from earlier in the function and doesn't need redeclaration. Changed `i` to `f` for clarity.

cilevitz committed:
Merge branch 'aws-deadline:mainline' into feature_font_support

Commit:
feat: Cache fontPaths during font scan
fix: Set submitConfirmation = true by default
docs: Warning on missing or substituted fonts
nit: Parameter spaces
nit: Variable names

Signed-off-by: Daniel Cilevitz <daniel@redtoast.ca>

cilevitz commented on dist/DeadlineCloudSubmitter.jsx:
```diff
@@ -1875,344 +1971,340 @@
+            "-" +
+            f(this.getUTCMonth() + 1) +
+            "-" +
+            f(this.getUTCDate()) +
```
This is a spurious error.

eherozhao commented on src/submission/JobTemplateHelper.jsx:
```diff
@@ -310,11 +343,35 @@ function getLocationForFont(fontPostScriptName) {
+            }
+        }
+    } catch (e) {
+        logger.error(e.message, jobTemplateHelperFile);
```
I saw this error logged in my log file.
> 2025-02-10 14:42:22 - [ERROR   ]  JobTemplateHelper.json: JSON.parse
2025-02-10 14:42:22 - [ERROR   ]  JobTemplateHelper.json: Object of type Object found where a Number, Array, or Property is needed

I think it was from this line. Can you please check if this was parsed correctly?

eherozhao reviewed (changes requested):
<img width="270" alt="image" src="https://github.com/user-attachments/assets/bbfe0fe1-37e0-40bc-92fb-5777b5043cef" />
I tried to submit a job for a scene with unsupported font. For the existing mainline version, the font was substituted by AE to some default font. However, using this PR branch, I didn't see the confirmation introduced from this PR asking about if I should submit the job if there is a missing or substituted fonts, also the job attachment was not having the substituted fonts. This will make the customer experience worse.

eherozhao commented on dist/DeadlineCloudSubmitter_Assets/JobTemplate/scripts/call_aerender.py:
```diff
@@ -91,6 +95,11 @@ def main():
+                # Some warnings should be treated as errors, i.e. a fatal warning
+                if any([w in line for w in FATAL_WARNINGS]):
+                    print(f"After Effects Fatal Warning: {line.strip()}", file=sys.stderr)
+                    sys.exit(1)
```
I submitted a job with missing fonts, but this didn't stop my job rendered. Here is the log.
```
2025/02/11 13:23:17-08:00 After Effects Warning: WARNING:After Effects warning: Grouped_Message_{This project contains references to missing effects. Please install the following effects to restore these references.|||This project contains a reference to a missing effect. Please install the following effect to restore this reference.###@0 missing effects.}"Saber"
2025/02/11 13:23:18-08:00 WARNING:After Effects warning: Project has missing fonts.
2025/02/11 13:23:18-08:00 After Effects Warning: WARNING:After Effects warning: Project has missing fonts.
2025/02/11 13:23:18-08:00  
2025/02/11 13:23:18-08:00 rqindex 1num 1
2025/02/11 13:23:18-08:00 PROGRESS:  2/11/2025 9:23:18 PM: Starting composition "test_comp".
2025/02/11 13:23:18-08:00  
2025/02/11 13:23:18-08:00 PROGRESS:  Render Settings: Best Settings
```

viknith commented:
Since there's been no updates or responses to comments and it's stale, we'll go ahead and close this PR. If it receives updates, go ahead and reopen the PR.
